File: shreyas_MATLAB_hole/war_simulator.m
%% description
% This script simulates a game of war!
%
% Authors: ChatGPT and Shreyas Kousik, but mostly ChatGPT
% Created: 2 Mar 2023

%% setup
% Generate a deck of cards represented as integers between 1 and 52
deck = randperm(52);

% play warrrr
playWar(deck) ;

%% how many turns do games of war usually take?
n_trials = 10000 ;
n_turns = nan(1,n_trials) ;
parfor idx = 1:n_trials
    deck = randperm(52) ;
    n_turns(idx) = playWar(deck) ;
end

histogram(n_turns')
xlabel('Number of Turns to Win')
ylabel('Number of Games')
make_plot_pretty()

%% play war
function turnCounter = playWar(deck)
    % Split the deck between two players
    player1Deck = deck(1:2:26);
    player2Deck = deck(2:2:end);
    
    % Initialize turn counter
    turnCounter = 0;
    
    % Play the game
    while ~isempty(player1Deck) && ~isempty(player2Deck) && turnCounter < 1000
        % Increase turn counter
        turnCounter = turnCounter + 1;
        
        % Draw a card from each player's deck
        player1Card = player1Deck(1);
        player1Deck = player1Deck(2:end);
        
        player2Card = player2Deck(1);
        player2Deck = player2Deck(2:end);
        
        % Take modulo 13 of the card values
        player1CardValue = mod(player1Card - 1, 13) + 1;
        player2CardValue = mod(player2Card - 1, 13) + 1;
        
        % Determine the winner of this round
        if player1CardValue > player2CardValue
            player1Deck = [player1Deck, player1Card, player2Card];
        elseif player2CardValue > player1CardValue
            player2Deck = [player2Deck, player1Card, player2Card];
        else
            [player1Deck, player2Deck] = war(player1Deck, player2Deck);
        end
    end
    
    % % Determine the winner of the game
    % if isempty(player1Deck)
    %     fprintf('Player 2 wins the game in %d turns!\n', turnCounter);
    % elseif isempty(player2Deck)
    %     fprintf('Player 1 wins the game in %d turns!\n', turnCounter);
    % else
    %     fprintf('Game terminated early!')
    % end
end

%% helper functions
function [deck1, deck2] = war(deck1, deck2)
    % Check if there are enough cards for war
    if length(deck1) < 4 || length(deck2) < 4
        % If one of the players doesn't have enough cards, the game ends
        deck1 = [];
        deck2 = [];
        return;
    end
    
    % Draw 3 cards from each player's deck
    player1Cards = deck1(1:3);
    deck1 = deck1(4:end);
    
    player2Cards = deck2(1:3);
    deck2 = deck2(4:end);
    
    % Draw a fourth card from each player's deck
    player1Card = deck1(1);
    deck1 = deck1(2:end);
    
    player2Card = deck2(1);
    deck2 = deck2(2:end);
    
    % Take modulo 13 of the card values
    player1CardValue = mod(player1Card - 1, 13) + 1;
    player2CardValue = mod(player2Card - 1, 13) + 1;
    
    % Determine the winner of this war
    if player1CardValue > player2CardValue
        deck1 = [deck1, player1Cards, player2Cards, player1Card, player2Card];
    elseif player2CardValue > player1CardValue
        deck2 = [deck2, player1Cards, player2Cards, player1Card, player2Card];
    else
        [deck1, deck2] = war(deck1, deck2);
    end
end

File: shreyas_MATLAB_hole/war_simulator.py
import random
import matplotlib.pyplot as plt

def play_war(deck):
    # Split the deck between two players
    player1_deck = deck[::2]
    player2_deck = deck[1::2]

    # Initialize turn counter
    turn_counter = 0

    # Play the game
    while player1_deck and player2_deck and turn_counter < 1000:
        # Increase turn counter
        turn_counter += 1

        # Draw a card from each player's deck
        player1_card = player1_deck.pop(0)
        player2_card = player2_deck.pop(0)

        # Take modulo 13 of the card values
        player1_card_value = (player1_card - 1) % 13 + 1
        player2_card_value = (player2_card - 1) % 13 + 1

        # Determine the winner of this round
        if player1_card_value > player2_card_value:
            player1_deck.extend([player1_card, player2_card])
        elif player2_card_value > player1_card_value:
            player2_deck.extend([player1_card, player2_card])
        else:
            player1_deck, player2_deck = war(player1_deck, player2_deck)

    # Return the number of turns taken
    return turn_counter

def war(deck1, deck2):
    # Check if there are enough cards for war
    if len(deck1) < 4 or len(deck2) < 4:
        # If one of the players doesn't have enough cards, the game ends
        return [], []

    # Draw 3 cards from each player's deck
    player1_cards = deck1[:3]
    deck1 = deck1[3:]

    player2_cards = deck2[:3]
    deck2 = deck2[3:]

    # Draw a fourth card from each player's deck
    player1_card = deck1.pop(0)
    player2_card = deck2.pop(0)

    # Take modulo 13 of the card values
    player1_card_value = (player1_card - 1) % 13 + 1
    player2_card_value = (player2_card - 1) % 13 + 1

    # Determine the winner of this war
    if player1_card_value > player2_card_value:
        deck1.extend(player1_cards + player2_cards + [player1_card, player2_card])
    elif player2_card_value > player1_card_value:
        deck2.extend(player1_cards + player2_cards + [player1_card, player2_card])
    else:
        deck1, deck2 = war(deck1, deck2)

    return deck1, deck2

# Generate a deck of cards represented as integers between 1 and 52
deck = list(range(1, 53))
random.shuffle(deck)

# Play a single game of war
turns_to_win = play_war(deck.copy())
print("Number of turns to win:", turns_to_win)

# # Simulate multiple games of war to estimate average number of turns
# n_trials = 10000
# n_turns = []
# for _ in range(n_trials):
#     deck = list(range(1, 53))
#     random.shuffle(deck)
#     n_turns.append(play_war(deck.copy()))
#
# # Plot histogram of number of turns
# plt.hist(n_turns, bins=range(1, max(n_turns) + 2), align='left', rwidth=0.8)
# plt.xlabel('Number of Turns to Win')
# plt.ylabel('Number of Games')
# plt.show()
